### PR TITLE
add: naive implementation of PPO_LLM

### PR DESCRIPTION
**Pull request by ss26**
additions:

**ppo llm**
- created `ppo_llm.py` closely following the BaseAgent structure that's currently used in `ppo.py`, using the notebook code from Jon and Nathaniel
- created `test_ppo_llm.py` to test basic functionality of each class in `ppo_llm.py`. all basic tests pass. 
    - to run these tests, check the usage command at the top of `test_ppo_llm.py`. run from the root directory. 
- added code to `train.py` for training using the new PPO LLM agent, HOWEVER i couldn't test the training because it exceeds my laptop's GPU memory :(

possible errors:
- since i couldn't train on my laptop, i'm unsure if the code works completely or if the ppo-llm agent will train at all, this is simply a starting point 

**import refactor**
- minor refactor of imports so running code from the `root_dir` can be centrally done using `python -m <exec command>` 
- updated the readme and added USAGE commands on the top of each relevant file

## Summary by Sourcery

Add a naive PPO_LLM implementation alongside import and CLI refactors to support language-model-based PPO training and evaluation from the project root.

New Features:
- Introduce PPO_LLM agent with ActorNetwork, CriticNetwork, and PPOLLMAgent classes for language-model-based PPO.
- Integrate PPO_LLM into train.py with custom trajectory collection, agent selection, and separate actor/critic checkpointing.

Enhancements:
- Refactor imports to use `src.` module paths and enable running scripts via `python -m` from the project root.

Documentation:
- Update README usage instructions to use `python -m src.<module>` syntax for training and evaluation commands.

Tests:
- Add comprehensive tests for ActorNetwork, CriticNetwork, and PPOLLMAgent covering forward passes, action selection, trajectory collection, GAEs, and update logic.

### CODE REVIEW

File: src/agent/ppo_llm.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        batch_size = len(actions)
+        states = [state] * batch_size
+        state_tokens = self.tokenizer(states, return_tensors="pt", padding=True)
+        action_tokens = self.tokenizer(
+            actions, return_tensors="pt", padding="longest", padding_side="right"
+        )
+        state_tokens = {k: v.to(device) for k, v in state_tokens.items()}
+        action_tokens = {k: v.to(device) for k, v in action_tokens.items()}
+        action_tokens["input_ids"] = action_tokens["input_ids"][:, 1:]  # Remove BOS
+        action_tokens["input_ids"] = torch.cat(
```
Comment (nmitrani) [inline src/agent/ppo_llm.py:27]: I don't understand why you would pad the tensor with zeros for each action to then set it at 1 (EOS). Isn't it simpler to pad it with torch.ones? Similarly, I don't understand when you add the indexing to the first dim of the action_tokens tensor [torch.arange(batch_size),...] Isn't the first dimension already of size batch_size, and could you not simply do: action_tokens[:,...]?

File: src/agent/ppo_llm.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            with torch.no_grad():
+                outputs = self.model(**inputs)
+        # Handle different model output types
+        if hasattr(outputs, 'logits'):
+            logits = outputs.logits
+        elif hasattr(outputs, 'last_hidden_state'):
+            # Add a linear head if not present
+            if not hasattr(self, 'lm_head'):
+                hidden_size = outputs.last_hidden_state.shape[-1]
+                self.lm_head = nn.Linear(hidden_size, self.tokenizer.vocab_size).to(device)
```
Comment (nmitrani) [inline src/agent/ppo_llm.py:61]: What models do not have a linear head? I had never heard or worked with any such model. If it's the case that they are common, good catch!

File: src/agent/ppo_llm.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+            logits = self.lm_head(outputs.last_hidden_state)
+        else:
+            raise AttributeError("Model output does not have 'logits' or 'last_hidden_state'.")
+        logits = torch.clamp(logits, min=-100, max=100)
+        k = state_tokens["input_ids"].shape[1]
+        m = action_tokens["input_ids"].shape[1]
+        next_token_logits = logits[:, k - 1 : -1, :]
```
Comment (nmitrani) [inline src/agent/ppo_llm.py:68]: Before you add an EOS, but here we don't count it towards the calculation right? The issue whether to include it or not is one we talked with Jon, so we can sync on that if you'd like, but for now it seems like you're adding it and then not taking it into account for computations

File: src/agent/ppo_llm.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        self.model = model
+        self.tokenizer = tokenizer
+        self.simple_nn = nn.Sequential(
+            nn.Flatten(),
+            nn.Linear(input_dim, 64),
+            nn.ReLU(),
+            nn.Linear(64, 64),
+            nn.ReLU(),
+            nn.Linear(64, 1),
+            nn.Sigmoid(),
```
[the remark this answers is not in the document]
Comment (nmitrani) [reply src/agent/ppo_llm.py:111]: The comment is correct, we should have a linear activation as final. Sigmoid restricts value estimations to [0,1], and without knowing the scale of rewards this is not necessarily sustainable. If we had a normalized reward in [0,1] then it would not be much of an issue. Right now, I suggest keeping a full ReLU NN.

File: src/agent/ppo_llm.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        )
+
+    def _set_device(self, device): 
+        self.model = self.model.to(device)
+        self.simple_nn = self.simple_nn.to(device)
+
+    def forward(self, state: str, device):
+        self._set_device(device)
+        assert self.model.device == device, f"Model device {self.model.device} does not match expected device {device}"
+        inputs = self.tokenizer(state, return_tensors="pt").to(device)
```
[the remark this answers is not in the document]
Comment (nmitrani) [reply src/agent/ppo_llm.py:121]: Check this, I don't know that it is true but it could very well be.

File: src/agent/ppo_llm.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            state = reset_result[0]
+        else:
+            state = reset_result
+        for _ in range(batch_size):
+            info = self.env._get_info()
+            available_actions = info['game_state']['choice_texts']
+            action, logprob, action_idx = self.choose_action(state, logprob_flag=True, return_idx=True)
+            next_state, reward, done, _ = self.env.step(action_idx)
+            trajectories.append((state, action, reward, done, next_state, logprob, available_actions))
+            state = next_state if not done else self.env.reset()
```
[the remark this answers is not in the document]
Comment (nmitrani) [reply src/agent/ppo_llm.py:200]: Agree

**Review by nmitrani — COMMENTED** [review]
Here is the review. Overall I think this was great work. Some stylistic changes, other things I was more confused about logic-wise: it could be just that I did not go deep enough to understand! Let me know and let's talk about it.

File: src/agent/ppo_llm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ss26) [reply src/agent/ppo_llm.py:121]: added a dict comprehension to move tensors to batch, as sourcery recommended

File: src/agent/ppo_llm.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    def forward(self, state: str, device):
+        self._set_device(device)
+        assert self.model.device == device, f"Model device {self.model.device} does not match expected device {device}"
+        inputs = self.tokenizer(state, return_tensors="pt").to(device)
+        with torch.no_grad():
+            outputs = self.model(**inputs, output_hidden_states=True)
+        # FOR TESTING ONLY: Use the first token's embedding from layer 1
+        hidden_states = outputs.hidden_states[1][:, 0, :].to(device)  # shape: [batch, hidden_size]
+        # hidden_states = outputs.hidden_states[1].to(device)  # shape: [batch, hidden_size]
+        x = self.simple_nn(hidden_states)
```
[the remark this answers is not in the document]
Comment (ss26) [reply src/agent/ppo_llm.py:127]: updated this and added an attention proj layer

File: src/agent/ppo_llm.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            delta = reward + self.gamma * next_value - value
+            gae = delta + self.gamma * self.lambda_gae * gae * (1 - done)
+            gaes.insert(0, gae)
+            returns.insert(0, gae + value)
+        gaes_tensor = torch.tensor(gaes, dtype=torch.float32)
+        returns_tensor = torch.tensor(returns, dtype=torch.float32)
+        if returns_flag:
+            return gaes_tensor, returns_tensor
+        else:
+            return gaes_tensor
```
[the remark this answers is not in the document]
Comment (ss26) [reply src/agent/ppo_llm.py:230]: cleaned up

File: tests/agent/test_ppo_llm.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def test_ppoagent_choose_action_machiavelli(gemma_resources, machiavelli_env):
+    device, tokenizer, model, config = gemma_resources
+    actor = ActorNetwork(model, tokenizer)
+    critic = CriticNetwork(model, tokenizer, config.hidden_size)
+    agent = PPOLLMAgent(machiavelli_env, actor, critic, device)
+    state, _ = machiavelli_env.reset()
+    action, logprob = agent.choose_action(state, logprob_flag=True)
+    assert isinstance(action, str) or isinstance(action, int)
```
[the remark this answers is not in the document]
Comment (ss26) [reply tests/agent/test_ppo_llm.py:87]: resolved

File: tests/agent/test_ppo_llm.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    critic = CriticNetwork(model, tokenizer, config.hidden_size)
+    agent = PPOLLMAgent(machiavelli_env, actor, critic, device)
+    batch_size = 3
+    trajectories = agent.collect_trajectories(batch_size)
+    assert len(trajectories) == batch_size
+    for t in trajectories:
+        state, action, reward, done, next_state, logprob, *_ = t
+        assert isinstance(state, str)
+        assert isinstance(action, str) or isinstance(action, int)
+        assert isinstance(reward, float) or isinstance(reward, int)
```
[the remark this answers is not in the document]
Comment (ss26) [reply tests/agent/test_ppo_llm.py:102]: resolved

File: src/agent/ppo_llm.py (hunk 3 above)
Comment (ss26) [reply src/agent/ppo_llm.py:68]: i'll discuss the EOS stuff with you on the call in greater detail. i ported the code from the notebook as-is, and i have some opinions about its place in the action tokens

File: src/agent/ppo_llm.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+            assert 'cuda' in str(self.device), f"Device should be CUDA, got {self.device}"
+        self.gamma = kwargs.get("gamma", 0.99)
+        self.clip_epsilon = kwargs.get("clip_epsilon", 0.1)
+        self.lambda_gae = kwargs.get("lambda_gae", 0.95)
+        # Remove duplicate parameters from actor and critic
+        params = list({id(p): p for p in list(self.actor.parameters()) + list(self.critic.parameters())}.values())
+        self.optimizer = torch.optim.Adam(
+            params,
+            lr=kwargs.get("learning_rate", 1e-5),
+        )
```
[the remark this answers is not in the document]
Comment (ss26) [reply src/agent/ppo_llm.py:158]: good catch. this is from an incorrect test i made before defining separate model heads for the actor and critic in `train.py`. fixed it.

File: tests/agent/test_ppo_llm.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+def test_gemma_actornetwork_conditional_logprobs(gemma_resources):
+    device, tokenizer, model, _ = gemma_resources
+    actor = ActorNetwork(model, tokenizer)
+    state = "Once upon a"
+    action1 = " time, there was a king."
+    action2 = " time, there was a princess."
+    actions = [action1, action2]
+    conditional_logprobs = actor.create_conditional_logprobs(state, actions, device)
+    print(conditional_logprobs)
+    torch.testing.assert_close(conditional_logprobs, torch.tensor([-0.1708, -1.8513], device=device), rtol=1e-3, atol=1e-3)
```
[the remark this answers is not in the document]
Comment (ss26) [reply tests/agent/test_ppo_llm.py:57]: removed brittle testing

File: src/agent/ppo_llm.py (hunk 3 above)
Comment (nmitrani) [reply src/agent/ppo_llm.py:61]: awesome, let's talk about it yes!

**Comment by nmitrani** [comment]
Mock models are actually a great idea, and one i was thinking we should implement as testing otherwise is very expensive. It could also help us test the integration with the Machiavelli env without loading models, which would be great.

**Review by nmitrani — APPROVED** [review]

**Comment by ss26** [comment]
i'm going to write ppo lagrangian in a branch `ppo_lag_llm_dev` that's checked out from this branch `ppo_llm_dev` assuming the next merge to the main branch will be this one #9 
informing so @domenicrosati can prioritize making this merge to avoid unnecessary rebasing later

**Comment by ss26** [comment]
- Updated `ppo_llm.py` to inherit the BaseAgent and its structure correctly (using act, observe, update respectively)
- Updated the testing to use MockModels to reduce testing load
